### PR TITLE
nixos/mycelium: init

### PR DESCRIPTION
**Pull request by Lassulus**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/networking/mycelium.nix [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+      type = lib.types.bool;
+      default = false;
+      description = "Open the firewall for mycelium";
+    };
+    package = lib.mkOption {
+      type = lib.types.package;
+      default = pkgs.mycelium;
+      description = "The mycelium package to use";
+    };
+    addHostedNodes = lib.mkOption {
```
Comment (zimbatm) [inline nixos/modules/services/networking/mycelium.nix:32]:
```suggestion
    addHostedPublicNodes = lib.mkOption {
```
A bit of a mouthful, but matches better with the upstream naming

File: nixos/modules/services/networking/mycelium.nix [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+let
+  cfg = config.services.mycelium;
+in
+{
+  options.services.mycelium = {
+    enable = lib.mkEnableOption "mycelium network";
+    peers = lib.mkOption {
+      type = lib.types.listOf lib.types.str;
+      description = "List of peers to connect to";
```
Comment (flokli) [inline nixos/modules/services/networking/mycelium.nix:11]: This needs a longer description / defaultText, to make it obvious the hosted peers are also added, at least it it's not explicitly disabled.

Comment (Lassulus) [reply nixos/modules/services/networking/mycelium.nix:11]: wouldn't that be better in the option where we add the hosted peers?

Comment (flokli) [reply nixos/modules/services/networking/mycelium.nix:11]: Just by looking at the "peers" config option, this strongly suggests there's no other peers.

The boostrap peers are not in that list, they're not discoverable via `config.` either.

What about my initial proposal, calling this `extraPeers`, and documenting it with "List of peers to connect to, in addition to these potentially enabled by `addHostedPublicNodes`"?

That should make it clear that these are not the only peers.

File: nixos/modules/services/networking/mycelium.nix [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      '';
+    };
+  };
+  config = lib.mkIf cfg.enable {
+    networking.firewall.allowedTCPPorts = lib.optionals cfg.openFirewall [ 9651 ];
+    networking.firewall.allowedUDPPorts = lib.optionals cfg.openFirewall [ 9650 9651 ];
+
+    systemd.services.mycelium = {
+      description = "Mycelium network";
+      wantedBy = [ "multi-user.target" ];
```
Comment (flokli) [inline nixos/modules/services/networking/mycelium.nix:50]: Please add a `After=network.target`, as per https://systemd.io/NETWORK_ONLINE/.

File: nixos/modules/services/networking/mycelium.nix [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+          "quic://[2a01:4f9:5a:1042::2]:9651"
+        ]));
+        Restart = "always";
+        RestartSec = 5;
+        TimeoutStopSec = 5;
+      };
+    };
+  };
+  meta = {
+    maintainers = with lib.maintainers; [ lassulus ];
```
Comment (flokli) [inline nixos/modules/services/networking/mycelium.nix:120]:
```suggestion
    maintainers = with lib.maintainers; [ flokli lassulus ];
```

File: nixos/modules/services/networking/mycelium.nix (hunk 2 above)
Comment (flokli) [reply nixos/modules/services/networking/mycelium.nix]: There's been a push at around the time I wrote this comment, now adding documentation, but keeping it `peers`, not `extraPeers`. Im not sure you're done. If so, please write a conclusion and mark the thread as resolved.

Comment (Lassulus) [reply nixos/modules/services/networking/mycelium.nix]: I still think peers is clearer. since extraPeers seems like peers but with more noise?

**Review by zimbatm — APPROVED** [review]

File: nixos/modules/services/networking/mycelium.nix [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        ExecStart = lib.concatStringsSep " " ([
+          (lib.getExe cfg.package)
+          (if (cfg.keyFile != null) then
+            "--key-file \${CREDENTIALS_DIRECTORY}/keyfile" else
+            "--key-file %S/mycelium/key"
+          )
+          "--tun-name" "mycelium"
+        ] ++
+          (lib.optional (cfg.addHostedPublicNodes || cfg.peers != []) "--peers")
+        ++ cfg.peers ++ (lib.optionals cfg.addHostedNodes [
```
Comment (flokli) [inline nixos/modules/services/networking/mycelium.nix:83]:
```suggestion
        ++ cfg.peers ++ (lib.optionals cfg.addHostedPublicNodes [
```

File: nixos/modules/services/networking/mycelium.nix [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    networking.firewall.allowedTCPPorts = lib.optionals cfg.openFirewall [ 9651 ];
+    networking.firewall.allowedUDPPorts = lib.optionals cfg.openFirewall [ 9650 9651 ];
+
+    systemd.services.mycelium = {
+      description = "Mycelium network";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      restartTriggers = [
+        cfg.keyFile
+      ];
```
Comment (flokli) [inline nixos/modules/services/networking/mycelium.nix:54]:
```suggestion
      ];
      
      unitConfig.Documentation = "https://github.com/threefoldtech/mycelium";
```

File: nixos/modules/services/networking/mycelium.nix [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        RestrictAddressFamilies = "AF_UNIX AF_INET AF_INET6 AF_NETLINK";
+        RestrictNamespaces = true;
+        RestrictRealtime = true;
+        SystemCallArchitectures = "native";
+        SystemCallFilter = [ "@system-service" "~@privileged @keyring" ];
+        ExecStart = lib.concatStringsSep " " ([
+          (lib.getExe cfg.package)
+          (if (cfg.keyFile != null) then
+            "--key-file \${CREDENTIALS_DIRECTORY}/keyfile" else
+            "--key-file %S/mycelium/key"
```
Comment (flokli) [inline nixos/modules/services/networking/mycelium.nix:78]:
```suggestion
            "--key-file %S/mycelium/key.bin"
```

That's the default according to their [example systemd unit](https://github.com/threefoldtech/mycelium/blob/f113c0b1e9c9c29c56c6486dae74013a46982e0e/systemd/mycelium.service#L15).

**Review by flokli — CHANGES_REQUESTED** [review]

**Review by flokli — APPROVED** [review]

**Pull request merged**